### ISSUE TITLE
FAVICONS results in unnecessary rebuilds

### ISSUE THREAD

**Issue by Kwpolska** [getnikola/nikola#1674]
via @chaghi
 I've found the problem... kind of... but can't explain it. It's definetely related to the FAVICONS set, when it has more than one tuple.

I copied my posts and pages to a clean site as suggested, and tried to reproduce the problem. I couldn't. I started to copy bit by bit other things (files, assets, the custom theme, etc.), and adjusting the bits of `conf.py` accordingly. At one point the "copied" site was like 90% of my real site, but still couldn't reproduce the problem. So I started to go through all the differences between my real `conf.py` and the "clean" `conf.py`, and eventually I was able to find out that the FAVICONS set was the culprit.

If I commented out the FAVICONS set, everything worked as expected. As soon as I uncommented the set, the bug was triggered. Then I found out that with just one tuple in the set I could not reproduce the problem either.

So I started a new test from scratch, to come up with a very minimal set of conditions to reproduce the problem 100% of the time... and I couldn't :( I found some situations in which the bug *might* be triggered. With a small-ish site you can run N consecutive builds without triggering the problem, and suddenly with the consecutive build N+1 (almost) all the html is regenerated. With a bigger site like mine (~500 posts), the unnecessary rebuild is triggered almost always. Sounds like another weird race condition, don't you think?

Anyway, after multiple tests both with my real site and minimal test sites created from scratch, I've found this:
- I could not trigger the bug with an empty FAVICONS set;
- I could not trigger the bug with a FAVICONS set with just 1 tuple;
- When the FAVICONS set had more than 1 tuple, the bug could be triggered almost always in large-ish sites, and sometimes with small-ish sites
- And for the fun part: I could NEVER trigger the bug if I turned the FAVICONS set in a tuple of tuples (like POSTS or PAGES), no matter how many tuples I put inside, and no matter the size of the site being built. Don't ask me why I tried this, even less why the type of the collection affects the result. :p Well, yes, sets are mutable, tuples are not, so there's that... but I can't explain it.

So, to sum this up: it would be nice if someone else tried this:
- Init a new site
- Add a couple of posts
- Add a couple of icons/images in the files folder

Then, for each one of this cases, do several consecutive `nikola build`:
- FAVICONS set empty/commented out
- FAVICONS set with just one tuple (pointing to one of the images)
- FAVICONS set with 2 tuples (pointing to both images)
- same as the previous case, but with FAVICONS defined as a tuple (replacing the "{}" by "()")

The size of the site seems to affect the result, so with a small site you might need to run several consecutive `nikola build` until the bug is triggered (i.e., most of the html is re-rendered even if you didn't touch anything in between).

Thanks!

**Comment by Kwpolska** [comment]
> set

Why the hell is `FAVICONS` a **set**?!  No wonder why it breaks, there are apparently issues with hashing (Python 3 uses a random order).

Make it a tuple in your config file, I’m fixing the defaults.  I’m not sure if we can fix this for all users by doing `sorted(self.config['FAVICONS'])`, there might be some special entry order needed… @Aeyoun?

cc @zhaojunmeng (via #256 63c6df2f7b5c5325224357d21f55b9c990406539)
